cmd/podman/machine/server9p.go

```diff
@@ -0,0 +1,101 @@
+//go:build windows && (amd64 || arm64)
+// +build windows
+// +build amd64 arm64
+
+package machine
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/containers/common/pkg/completion"
+	"github.com/containers/podman/v4/cmd/podman/registry"
+	"github.com/containers/podman/v4/pkg/fileserver"
+	psutil "github.com/shirou/gopsutil/v3/process"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+)
+
+var (
+	server9pCommand = &cobra.Command{
+		Args:              cobra.ExactArgs(1),
+		Use:               "server9p [options] PID",
+		Hidden:            true,
+		Short:             "Serve a directory using 9p over hvsock",
+		Long:              "Start a number of 9p servers on given hvsock UUIDs, and run until the given PID exits",
+		RunE:              remoteDirServer,
+		ValidArgsFunction: completion.AutocompleteNone,
+		Example:           `podman system server9p --serve C:\Users\myuser:00000050-FACB-11E6-BD58-64006A7986D3 /mnt`,
+	}
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: server9pCommand,
+		Parent:  machineCmd,
+	})
+
+	flags := server9pCommand.Flags()
+
+	serveFlagName := "serve"
+	flags.StringArrayVar(&serveDirs, serveFlagName, []string{}, "directories to serve and UUID of vsock to serve on, colon-separated")
+	_ = server9pCommand.RegisterFlagCompletionFunc(serveFlagName, completion.AutocompleteNone)
+}
+
+var (
+	serveDirs []string
+)
+
+func remoteDirServer(cmd *cobra.Command, args []string) error {
+	pid, err := strconv.Atoi(args[0])
+	if err != nil {
+		return fmt.Errorf("parsing PID: %w", err)
+	}
+	if pid < 0 {
+		return fmt.Errorf("PIDs cannot be negative")
+	}
+
+	if len(serveDirs) == 0 {
+		return fmt.Errorf("must provide at least one directory to serve")
+	}
+
+	// TODO: need to support options here
+	shares := make(map[string]string, len(serveDirs))
+	for _, share := range serveDirs {
+		splitShare := strings.Split(share, ":")
+		if len(splitShare) < 2 {
+			return fmt.Errorf("paths passed to --share must include an hvsock GUID")
+		}
+
+		// Every element but the last one is the real filepath to share
+		path := strings.Join(splitShare[:len(splitShare)-1], ":")
+
+		shares[path] = splitShare[len(splitShare)-1]
+	}
+
+	if err := fileserver.StartShares(shares); err != nil {
+		return err
+	}
+
+	p, err := psutil.NewProcess(int32(pid))
+	if err != nil {
+		return fmt.Errorf("opening gvproxy PID: %w", err)
+	}
+	for {
+		running, err := p.IsRunning()
+		if err != nil {
+			return fmt.Errorf("checking is gvproxy is dead: %w", err)
+		}
+		if !running {
+			break
+		}
+
+		time.Sleep(1 * time.Second)
+	}
```

Comment:
I am not quite sure why this is needed? There seems to be a WaitForExit() call in windows which unlike linux should also work if the process you are waiting for is not a child. Are you something like https://stackoverflow.com/questions/37664015/os-process-wait-after-os-findprocesspid-works-on-windows-not-on-linux does not work and we really need such a big dep?

Alternative why can't we just use the same logic as for gvproxy stop and let the machine commands kill this process and let it sleep forever here instead of this busy loop.

Reply:
The WaitForExit call is not accessible via Golang's standard library without DLL loading shenanigans and then calling untyped functions (well, the function itself is, but the ability to get the handle that it requires to work is not) - so this is *a lot* cleaner, even if it's technically gross.

I don't have objections to using a PID file and terminating on machine stop, though. @baude You have strong feelings on this?

Reply:
Further, I'm strongly in favor of keeping the dep even if we drop this code. The things we were doing with Go's native `os.Process` code to do things like check for process exit just did not work - I solved 2 separate bugs elsewhere in the code by replacing our custom-rolled Windows/Linux process code with this library.

Reply:
I you think it is needed sure, not like I tested it on windows. I just cannot believe that the go std lib wouldn't suffice for that use case but if it is like that then sure.

Reply:
From what I can see, the standard library is fine on basically any POSIX-compliant system, but it starts to fall apart on Windows - things are sufficiently different that the abstractions can't handle it.

Reply:
I agree here, the stdlib on Windows does not provide the same functionality as you have on Posix-compliant systems. I think this is expectable for now, as we might spin this code out at a later time.